### PR TITLE
Rewrite scaling Vespa

### PR DESCRIPTION
**Pull request by jobergum**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by jobergum** [comment]
First cut at improving sizing documentation, please review but do not merge yet.

File: documentation/attributes.html [hunk 1]
```diff
@@ -62,44 +62,27 @@ <h2 id="search">Search</h2>
...
-Hence, increasing an attribute means more than doubling the memory used -
-refer to the <a href="content/setup-proton-tuning.html#resizing">resizing reference</a>.
-</p><p>
-<a href="search-definitions.html#multivalue-fields">Multivalue</a> attributes use an adaptive approach in how data is stored in memory,
-and up to 2 billion documents per node is supported.
-</p><p>
-Read the <a href="performance/attribute-memory-usage.html">attribute sizing guide</a>.
+can update memory structures with high throughput and low latency without having to read the original copy from the document store.
+The Vespa search core maintains separate data structures for the document <a href="proton.html#sub-databases">sub databases</a>
+, where only the ready sub database is in memory one should consider making also the not ready sub database attribute data in memory by
```
Comment (kkraune) [inline documentation/attributes.html]:
```suggestion
, where only the <em>ready</em> sub database is in memory one should consider making also the <em>not ready</em> sub database attribute data in memory by 
```

File: documentation/attributes.html [hunk 2]
```diff
@@ -62,44 +62,27 @@ <h2 id="search">Search</h2>
...
-</p><p>
-Read the <a href="performance/attribute-memory-usage.html">attribute sizing guide</a>.
+can update memory structures with high throughput and low latency without having to read the original copy from the document store.
+The Vespa search core maintains separate data structures for the document <a href="proton.html#sub-databases">sub databases</a>
+, where only the ready sub database is in memory one should consider making also the not ready sub database attribute data in memory by 
+configuring the attribute 
+<a href="reference/search-definitions-reference.html#attribute">fast-access</a>. 
 </p>
 
+<<<<<<< HEAD
```
Comment (kkraune) [inline documentation/attributes.html]: ?

File: documentation/content/setup-proton-tuning.html [hunk 3]
```diff
@@ -187,8 +187,9 @@ <h2 id="requestthreads">requestthreads</h2>
 Tune the number of request threads used on a search node - optional sub-elements:
 <ul>
   <li id="requestthreads-search"><code>search</code>: Number of search threads, default 64</li>
-  <li id="requestthreads-persearch"><code>persearch</code>: Number of search threads used per search,
-    default 1</li>
+  <li id="requestthreads-persearch"><code>persearch</code>: Number of search threads used per search, see <a href="../performance/sizing-search.html">Sizing Vespa Search</a> documentation
```
Comment (kkraune) [inline documentation/content/setup-proton-tuning.html]: linebreak

File: documentation/performance/attribute-memory-usage.html [hunk 4]
```diff
@@ -60,7 +60,14 @@ <h3 id="example">Example</h3>
 with various data types and collection types.
 It also contains estimates for how many documents a 48GB RAM node can hold, taking initialization into account.
 </p>
-
+<p>
+When <em>resizing</em> attributes, both current and new attribute will temporarily reside in memory.
+Hence, increasing an attribute means more than doubling the memory used -
```
Comment (kkraune) [inline documentation/performance/attribute-memory-usage.html]: I thought this was changed, at least for multivalue attributes? Ref Geir/Tor work last Q

File: documentation/performance/feature-tuning.html [hunk 5]
```diff
@@ -1,39 +1,72 @@
 ---
 # Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
-title: "Vespa Search Performance Tuning"
+title: "Vespa Serving Tuning"
 ---
 <p>
-This document describes how to tune an application for high performance,
-while <a href="sizing-search.html">search sizing guide</a> is about scaling an application.
+This document describes tuning certain features of an application for high performance, this document mainly focus on content cluster search features, see <a href="container-tuning.html">Container tuning</a> for tuning of container clusters</a>.
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]: linebreaks

File: documentation/performance/feature-tuning.html [hunk 6]
```diff
@@ -1,39 +1,72 @@
...
+See <a href="sizing-search.html">search sizing guide</a> is about scaling an application deployment.
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]:
```suggestion
The <a href="sizing-search.html">search sizing guide</a> is about scaling an application deployment.
```

File: documentation/performance/feature-tuning.html [hunk 7]
```diff
@@ -1,39 +1,72 @@
...
 </p>
 
-
-
 <h2 id="attribute-vs-index">Attribute v.s index</h2>
 <p>
-The <a href="../attributes.html">attribute</a> documentation summaries when to use <samp>attribute</samp>
+The <a href="../attributes.html">attribute</a> documentation summaries when to use <a href="../reference/search-definitions-reference.html#attribute">attribute</a>
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]:
```suggestion
The <a href="../attributes.html">attribute</a> documentation documents when to use <a href="../reference/search-definitions-reference.html#attribute">attribute</a> 
```

File: documentation/performance/feature-tuning.html [hunk 8]
```diff
@@ -1,39 +1,72 @@
...
 <pre>
 field timestamp type long {
-    indexing:  summary | attribute
+    indexing: summary | attribute
+    rank: filter
+}
+</pre>
+
+<p>If both index and attribute is configured for string type fields,
+Vespa will do search and matching against the index with default match <samp>text</samp>. All numeric type fields and tensor fields are attribute (in-memory)
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]: linebreak after .  - makes later edits easier ...

File: documentation/performance/feature-tuning.html [hunk 9]
```diff
@@ -1,39 +1,72 @@
...
+fields in Vespa.</p>
+
+<h2 id="when-to-use-fast-search">When to use fast-search for attribute fields</h2>
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]:
```suggestion
<h2 id="when-to-use-fast-search-for-attribute-fields">When to use fast-search for attribute fields</h2>
```

File: documentation/performance/feature-tuning.html [hunk 10]
```diff
@@ -1,39 +1,72 @@
...
+<p> By default there Vespa does not build any posting list index structures over <em>attribute</em> fields, adding <em>fast-search</em> to the attribute
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]:
```suggestion
<p> By default, Vespa does not build any posting list index structures over <em>attribute</em> fields. Adding <em>fast-search</em> to the attribute 
```

File: documentation/performance/feature-tuning.html [hunk 11]
```diff
@@ -1,39 +1,72 @@
...
 </pre>
-If both index and attribute is configured for string type fields,
-Vespa will do search and matching against the index with default match <samp>text</samp>.
-</p>
+<p>When Vespa executes a query with multiple query items it builds an query execution plan which it tries to optimize so that query tree items that 
+are restrictive meaning they match few documents are evaluated early so that the temporary result set becomes as low as possible. To do so the query execution 
+plan looks at hit count estimates for each part of the query tree using the index and b-tree dictionaries which track the number of documents a given term occurs in. However, for
+attribute fields without fast-search there is no hit count estimate so the estimate becomes equal to the total number of documents (matches all) and is thus moved
+to the end of the query evaluation. A query with only one query term searching a attribute field without <em>fast-search</em> would be a linear scan over all documents
+and thus very slow:
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]: rewrite to shorter sentences, also using a comma here and there :-)

File: documentation/performance/feature-tuning.html [hunk 12]
```diff
@@ -1,39 +1,72 @@
...
+select * from sources * where range(timestamp, 0, 100) and uuid contains "123e4567-e89b-12d3-a456-426655440000";
+</pre>
 
+The general rules of thumb for when to use fast-search for an attribute field is: 
+
+<ul>
+  <li>Use <em>fast-search</em> if the attribute field is searched without any other query terms</li>
+  <li>Use <em>fast-search</em> if the attribute field could limit the total number of hits efficiently</li>
+</ul>
+<p>Changing fast-search aspect of the attribute is a live change which does not require any re-feeding so testing the performance with and without is low effort</p>
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]: break up - tooooo long sentence

File: documentation/performance/feature-tuning.html [hunk 13]
```diff
@@ -94,29 +127,25 @@ <h2 id="parent-child-and-search-performance">Parent child and search performance
 </p>
 
 
-
-<h2 id="ranking">Ranking</h2>
+<h2 id="ranking">Ranking and ML Model inferences </h2>
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]: id and text shall be the same

File: documentation/performance/feature-tuning.html [hunk 14]
```diff
@@ -94,29 +127,25 @@ <h2 id="parent-child-and-search-performance">Parent child and search performance
...
-<a href="sizing-search.html">Vespa scales</a> with the number of hits the query recalls
-which needs to be ranked per node.
-The ranking cost per document recalled is determined by the complexity of the ranking expression in use
-and the rank feature complexity.
-</p>
-
+<a href="sizing-search.html">Vespa scales</a> with the number of hits the query retrieves per node/search thread and which needs to be evaluated
+by the first-phase ranking function. Read more on <a href="../phased-ranking.html">phased ranking</a>, phased ranking enables spending more resources 
+during a second phase ranking step than in the first-phase, the first-phase is about getting a decent recall (retrieve relevant documents in the top k) and second phase is
+used to tune precision at k, bringing the relevant documents to the top of the result set.
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]: split

File: documentation/performance/feature-tuning.html [hunk 15]
```diff
@@ -94,29 +127,25 @@ <h2 id="parent-child-and-search-performance">Parent child and search performance
...
-The ranking cost per document recalled is determined by the complexity of the ranking expression in use
-and the rank feature complexity.
-</p>
-
+<a href="sizing-search.html">Vespa scales</a> with the number of hits the query retrieves per node/search thread and which needs to be evaluated
+by the first-phase ranking function. Read more on <a href="../phased-ranking.html">phased ranking</a>, phased ranking enables spending more resources 
+during a second phase ranking step than in the first-phase, the first-phase is about getting a decent recall (retrieve relevant documents in the top k) and second phase is
+used to tune precision at k, bringing the relevant documents to the top of the result set. 
 
+. For text ranking applications one might consider using the <a href="../using-wand-with-vespa.html">WAND</a>
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]:
```suggestion
For text ranking applications one might consider using the <a href="../using-wand-with-vespa.html">WAND</a>
```

File: documentation/performance/feature-tuning.html [hunk 16]
```diff
@@ -94,29 +127,25 @@ <h2 id="parent-child-and-search-performance">Parent child and search performance
...
+query operator which can efficiently find the top k documents using a internal linear ranking function.
+</p>
 
-<h2 id="document-summaries">Document summaries</h2>
+<h2 id="document-summaries">Document summaries - hits</h2>
```
Comment (kkraune) [inline documentation/performance/feature-tuning.html]:
```suggestion
<h2 id="document-summaries-hits">Document summaries - hits</h2>
```

File: documentation/performance/sizing-examples.html [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+
+
+
+
+
+
+ 
+  
+
+</p>
```
Comment (kkraune) [inline documentation/performance/sizing-examples.html]: I will review this after merge

File: documentation/performance/sizing-search.html [hunk 18]
```diff
@@ -1,705 +1,477 @@
...
-    com.yahoo.prelude.querytransform.LiteralBoostSearcher@music(QueryProcessing(SEARCH: 0 ms), ResultProcessing(SEARCH: 0 ms)),
-    com.yahoo.prelude.querytransform.CJKSearcher@music(QueryProcessing(SEARCH: 0 ms), ResultProcessing(SEARCH: 0 ms)),
-    com.yahoo.prelude.querytransform.StemmingSearcher@music(QueryProcessing(SEARCH: 0 ms), ResultProcessing(SEARCH: 0 ms)),
-    com.yahoo.prelude.querytransform.NormalizingSearcher@music(QueryProcessing(SEARCH: 1 ms), ResultProcessing(SEARCH: 0 ms)),
-    com.yahoo.prelude.searcher.ValidateSortingSearcher@music(QueryProcessing(SEARCH: 0 ms), ResultProcessing(SEARCH: 0 ms)),
-    com.yahoo.prelude.cluster.ClusterSearcher@music(QueryProcessing(SEARCH: 0 ms), ResultProcessing()),
-    <span style="background-color: yellow;">com.yahoo.search.SlowSearcher@default(QueryProcessing(SEARCH: 12 ms, FILL: 0 ms), ResultProcessing(SEARCH: 18 ms)),</span>
-    com.yahoo.prelude.statistics.StatisticsSearcher@native(QueryProcessing(SEARCH: 1 ms), ResultProcessing(SEARCH: 0 ms)),
-    com.yahoo.search.grouping.GroupingQueryParser@v
```
Comment (kkraune) [inline documentation/performance/sizing-search.html]: will review after merge

File: documentation/performance/attribute-memory-usage.html (hunk 4 above)
Comment (jobergum) [reply documentation/performance/attribute-memory-usage.html]: Yes, removed it. I

File: documentation/performance/sizing-search.html [hunk 19]
```diff
@@ -1,705 +1,477 @@
...
-See the graph below, where we have plotted the 99 percentile latency of a fixed set of
-queries executed against an increasing number of documents indexed.
+When <em>SQW</em> is no longer significantly less than <em>DQW</em>,
+adding more nodes in a flat distributed cluster just increases the overall system cost without any serving performance gain except increasing overall 
+supported feed throughput which will increase almost linearly with number of nodes. 
+
+Two different <em>DQW/(DQW+SQW)</em> factors are  
+are illustrated in the figures below. The overall query work (TQW) is the same for both cases (10 ms) but the <em>DQW</em> portion of the <em>TQW</em> is different. 
+The throughput (QPS) is 
+a function of the latency (<a href="https://en.wikipedia.org/wiki/Little%27s_law">Little's Law</a>) and the number of cpu cores * nodes. Using 1 node with 24 v-cpu cores and 10 ms service time (<em>TQW</em> we expect to reach close to 2400 QPS at 100% utilization (Unless there are other bottlenecks downstream).
```
Comment (geirst) [inline documentation/performance/sizing-search.html]: Missing ")" after TQW

File: documentation/performance/sizing-search.html [hunk 20]
```diff
@@ -1,705 +1,477 @@
...
+Two different <em>DQW/(DQW+SQW)</em> factors are  
+are illustrated in the figures below. The overall query work (TQW) is the same for both cases (10 ms) but the <em>DQW</em> portion of the <em>TQW</em> is different. 
+The throughput (QPS) is 
+a function of the latency (<a href="https://en.wikipedia.org/wiki/Little%27s_law">Little's Law</a>) and the number of cpu cores * nodes. Using 1 node with 24 v-cpu cores and 10 ms service time (<em>TQW</em> we expect to reach close to 2400 QPS at 100% utilization (Unless there are other bottlenecks downstream). 
 </p>
+<figure>
+  <img src="../img/ScalingLatencyFactor0.5.svg" />
+  <img src="../img/ScalingLatencyFactor0.005.svg" />
+</figure>
+In the left figure the overall latency is 10 ms but the dynamic query work (latency) is only 50% and from Amdahl's law it follows that the maximum speedup we can get is 2 regardless how many
```
Comment (geirst) [inline documentation/performance/sizing-search.html]: The figures are rendered on top of each other if width of page is low, consider naming them as first and second.

File: documentation/performance/sizing-search.html [hunk 21]
```diff
@@ -1,705 +1,581 @@
...
+    </tr><tr id="searchable-copies">
+      <th>Searchable copies</th>
+      <td>
+        <p>
+        <a href="../reference/services-content.html#searchable-copies">Searchable-copies</a>
+        configures how many of the copies (as configured with <em>redundancy</em>) to be indexed 
+        (<em>ready</em>) at any time. Configuring
+        <em>searchable-copies</em> to be less than redundancy saves resources (memory, disk, cpu) 
+        as not all copies are indexed (<em>ready</em>).  In case 
+        of node failure the remaining nodes needs to activate (index) the
```
Comment (vekterli) [inline documentation/performance/sizing-search.html]: Technically, activating documents is a separate operation from indexing them (we activate ready replicas to un-blacklist them), so may consider just rewriting to "(...) the remaining nodes need to index the (...)"

File: documentation/performance/sizing-search.html [hunk 22]
```diff
@@ -1,705 +1,477 @@
...
+      <th style="white-space:nowrap;">Add groups using grouped distribution</th>
+      <td>When DQW/TQW is low, one can no longer just add more content
+      nodes to scale throughput and must instead use a grouped distribution to scale throughput. 
+    </tr>
+  </tbody>
+</table>
+
+<h2 id="scaling-latency">Scaling latency in a content group</h2>
+<p>
+Wether we have a single group (flat distribution) or multiple groups, the serving latency depends on the factors already described; DQW and SQW. For use cases where DQW
```
Comment (geirst) [inline documentation/performance/sizing-search.html]: Wether -> Whether

File: documentation/performance/sizing-search.html [hunk 23]
```diff
@@ -1,705 +1,581 @@
...
+permanently deleted
+from storage by <a href="proton.html#proton-maintenance-jobs">Proton maintenance jobs</a></li>
+
+</ul>
+<p>If the availability guarantees tolerate temporary search coverage loss during node 
+failures using <em>searchable-copies</em> 1 is by far the 
+most optimal for for serving performance as query work per node becomes
+less as index structures like posting lists are smaller as they then only contains documents in <em>Active</em> 
+state and not also the <em>Not Active</em> documents.</p>
+With <em>searchable-copies</em> 2 and <em>redundancy</em> 2 each content node will ready index both replicas
```
Comment (vekterli) [inline documentation/performance/sizing-search.html]: "each content node" may be a bit confusing here, consider perhaps rephrasing as `With <em>searchable-copies</em> 2 and <em>redundancy</em> 2 each replica is fully indexed on separate content nodes (...)`

File: documentation/performance/sizing-search.html [hunk 24]
```diff
@@ -1,705 +1,477 @@
...
 </p>
 
+<h2 id="scaling-document-volume">Scaling document volume per node</h2>
+<p>
+One want to fit as many documents as possible into a node given the node flavor constrains (e.g available cpu, memory, disk) while maintaining 
+<ul>
+  <li>Our targeted search latency SLA</li>
+  <li>Our targeted feed and update rate</li>
+</ul> 
+With the latency SLA in mind we should benchmark with increasing number of documents per node and watch system level metrics and Vespa metrics. If latency is within SLA and sustained feed rate
```
Comment (geirst) [inline documentation/performance/sizing-search.html]: Word(s) missing before "sustained feed rate"? Consider rewrite.

File: documentation/performance/sizing-search.html [hunk 25]
```diff
@@ -1,705 +1,477 @@
...
+<h2 id="scaling-document-volume">Scaling document volume per node</h2>
+<p>
+One want to fit as many documents as possible into a node given the node flavor constrains (e.g available cpu, memory, disk) while maintaining 
+<ul>
+  <li>Our targeted search latency SLA</li>
+  <li>Our targeted feed and update rate</li>
+</ul> 
+With the latency SLA in mind we should benchmark with increasing number of documents per node and watch system level metrics and Vespa metrics. If latency is within SLA and sustained feed rate 
+we reduce the overall cost by fitting more documents in a given node (e.g increasing memory, cpu and disk constraints). Vespa will block feed operations
+if <a href="../reference/services-content.html#resource-limits">resource limits</a> has been reached.</p>
```
Comment (geirst) [inline documentation/performance/sizing-search.html]: has -> have

File: documentation/performance/sizing-search.html [hunk 26]
```diff
@@ -1,705 +1,477 @@
...
-Do note that in a default vespa setup there will be multiple other processes
-running along with the vespa-proton process that will also consume memory.
+so the virtual memory usage reported might be very high as both index and summary files are mapped into memory using mmap and pages are paged into memory as needed.
 </p><p>
 The memory usage per document depends on the number of fields,
-the size of the document and how many of the fields are defined as attributes.
-</p>
+the size of the documents and how many of the fields are defined as <a href="https://docs.vespa.ai/documentation/attributes.html">attributes</a>.
+See <a href="attribute-memory-usage.html">attribute memory usage sizing</a>
+and See <a href="#metrics-for-capacity-plan">metrics for capacity planning</a>.
```
Comment (geirst) [inline documentation/performance/sizing-search.html]: See -> see

File: documentation/performance/sizing-search.html [hunk 27]
```diff
@@ -1,705 +1,477 @@
...
 
 
+<h2 id="scaling-throughput">Scaling Throughput</h2>
+<p>
+As seen in the previous sections When the static query work (<em>SQW</em>) becomes large we should scale throughput by using grouped distribution but it should also be noted that scaling
```
Comment (geirst) [inline documentation/performance/sizing-search.html]: "sections When" -> "sections, when"

File: documentation/performance/sizing-search.html [hunk 28]
```diff
@@ -1,705 +1,477 @@
...
-</p>
+the size of the documents and how many of the fields are defined as <a href="https://docs.vespa.ai/documentation/attributes.html">attributes</a>.
+See <a href="attribute-memory-usage.html">attribute memory usage sizing</a>
+and See <a href="#metrics-for-capacity-plan">metrics for capacity planning</a>.
 
 
+<h2 id="scaling-throughput">Scaling Throughput</h2>
+<p>
+As seen in the previous sections When the static query work (<em>SQW</em>) becomes large we should scale throughput by using grouped distribution but it should also be noted that scaling 
+throughput by group replication increases storage and memory requirements. Regardless if we have scaled throughput by grouped distribution for use cases with high static query work portion or
```
Comment (geirst) [inline documentation/performance/sizing-search.html]: "Regardless..." sentence is long and hard to grasp. Consider simplifications.

File: documentation/performance/sizing-search.html [hunk 29]
```diff
@@ -1,705 +1,477 @@
...
 At some throughput level, some resource(s) in the system will be
-saturated and requests will be queued up causing latency to spike up
-exponentially since requests are arriving faster then they they are able to be served.
-The more queries executed past the saturation point,
-the longer an average request needs to be in the queue waiting to be served.
-This is demonstrated in the graph below for the 10M documents per node case:
+fully saturated and requests will be queued up causing latency to spike up
+exponentially since requests are spending more time in the queue.  
+The more queries/s we try to push by increasing the load 
+, the longer request needs to be in the queue waiting to be served. This behaviour is illustrated in the figure below.
```
Comment (geirst) [inline documentation/performance/sizing-search.html]: "request" -> "a request"

File: documentation/reference/services-content.html [hunk 30]
```diff
@@ -417,20 +417,9 @@ <h2 id="group">group</h2>
...
-  &lt;group name="bottom-1" distribution-key="0">
-    &lt;node distribution-key="0" hostalias="node1"/>
-  &lt;/group>
-  &lt;group name="bottom-2" distribution-key="1">
-    &lt;node distribution-key="1" hostalias="node2"/>
-  &lt;/group>
-&lt;/group>
-</pre>
-</p>
+</p><p>See <a href="../performance/sizing-search.html">Scaling Vespa Serving</a> for when to consider using
```
Comment (geirst) [inline documentation/reference/services-content.html]: "Scaling Vespa Serving" -> "Vespa Serving Scaling Guide"

File: documentation/reference/services-content.html [hunk 31]
```diff
@@ -999,9 +914,10 @@ <h2 id="dispatch-tuning">dispatch</h2>
         Declares the maximum number of hits to return from a single node.
         By default, a query returns the requested number of hits +
         offset from every search node up to the container,
-        which in turns orders them according to the query,
+        which in turns orders them globally according to the query,
         then discards all hits beyond the number requested.
-        In a system with a large fan-out, this can consume a lot of bandwidth.
+        In a system with a large fan-out, this can consume a lot of network bandwidth and the container nodes
+        are easily network saturated even if the query is performed in. See <a href="../performance/sizing-search.html">Vespa Search Sizing</a>.
```
Comment (geirst) [inline documentation/reference/services-content.html]: Missing word(s) at end of sentence "...performed in."?

File: documentation/reference/services-content.html [hunk 32]
```diff
@@ -1013,16 +929,18 @@ <h2 id="dispatch-tuning">dispatch</h2>
       <td>round-robin / adaptive</td>
       <td>round-robin</td>
       <td>
-        Configure policy for choosing which <a href="../qps-scaling-using-groups.html">group</a>
-        shall receive the next request. However multi-phase
+        Configure policy for choosing which group 
+        shall receive the next query request.
+        However multi-phase
         requests that either requires or benefits from hitting the same group in all phases are always hashed.
+        Relevant only for <a href="../performance/sizing-search.html">grouped distribution</a>.
```
Comment (geirst) [inline documentation/reference/services-content.html]: Consider using "#data-distribution" in link

File: documentation/reference/services-content.html [hunk 33]
```diff
@@ -1031,7 +949,7 @@ <h2 id="dispatch-tuning">dispatch</h2>
       <td>100</td>
       <td>
         Coverage required in order to serve from a group - default full coverage.
-        Relevant only for <a href="../qps-scaling-using-groups.html">grouped distribution</a>.</td></tr>
+        Relevant only for <a href="../performance/sizing-search.html">grouped distribution</a>.</td></tr>
```
Comment (geirst) [inline documentation/reference/services-content.html]: Consider using "#data-distribution" in link

File: documentation/reference/services-content.html [hunk 34]
```diff
@@ -1040,7 +958,7 @@ <h2 id="dispatch-tuning">dispatch</h2>
         Percentage of active documents a group needs to have
         compared to average of other groups in order to be active for serving queries.
         Because of measurement timing differences, it is not advisable to tune this above 99 percent.
-        Relevant only for <a href="../qps-scaling-using-groups.html">grouped distribution</a>.</td></tr>
+        Relevant only for <a href="../performance/sizing-search.html">grouped distribution</a>.</td></tr>
```
Comment (geirst) [inline documentation/reference/services-content.html]: Consider using "#data-distribution" in link

File: documentation/performance/attribute-memory-usage.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (jobergum) [reply documentation/performance/attribute-memory-usage.html]: Thanks, have removed.

**Pull request merged**